DMTN-148.tex

```diff
@@ -50,9 +50,255 @@
 % ADD CONTENT HERE
 % You can also use the \input command to include several content files.
 
+\section{Intro}
+
+The goal of this tech note is to provide a plan for the handling of
+calibration products in the Gen-3 Butler environment.  There are a
+large number of these products currently in use, but many of them do
+not have a clearly defined process to define, ingest, test, and
+deprecate them.  The lack of documentation on how these processes work
+has led to many implementations of similar paths, further complicating
+the issue.
+
+
+\section{The End State}
+
+The final state that we wish to get to has the following properties:
+
+\begin{itemize}
+\item All calibration products are either lsst.afw.image.Exposure
+  image data or an explicit calibration class in the \verb|ip_isr|
+  package.\footnote{This will require class reorganization to migrate
+    the Defects and Curve calibrations definitions.}
+\item All calibration products should be constructed from a DM Task in
+  the \verb|cp_pipe| package, and have the input data and
+  configuration parameters persisted with the calibration itself.
+  This guarantees that any product can be reconstructed in the future.
+  Exceptions will exist for external or lab created calibrations,
+  although the majority of used calibrations will likely be produced
+  directly.
+\item These provenance, config, and final product files may be saved
+  to an \verb|obs_CAMERA_data| package in git for convenience.  This also allows
+  these files to be migrated to other Butler repositories in a
+  consistent fashion.
+\item All calibration products are accepted for use and have a date
+  range set manually, preventing test runs from being used
+  inadvertently.  Calibrations ingested from git should have sufficient
```

Comment:
I am no astronomer, so forgive the naïveté, but... this logic always surprises me.

It seems like you might want to select calibration products in a couple of different ways. The first is well supported by the above: given some observation date, I can work out which calibration products are applicable by looking at their date range. Fine.

However, it also seems like I'd want to be able to select calib products over specific ranges of time: given some observation date, and what I know about the instrument, I _choose_ to accept calib products generated within _n_ time-periods of the data being taken.

Of course, I can still do that query manually based on `CALIBDATE`... but if the only way of knowing if the product actually originates from a test run is to try to infer that based on its `VALIDSTART`/`VALIDEND`, that feels really fragile. (If the difference is 0, fine... but what if it's one second? One minute? Five minutes?).

Maybe this is just something that we never want to be able to do...?

Reply:
This needs more detail about the gen3 certification process, which is still not finalized.  However, a test run will never be visible outside of the gen3 collection it was constructed in.  This test run will have no `VALIDSTART/END` value, because it hasn't been validated and certified.  There will be (but it has not been added), some way to ensure validation processing can still work from this collection, but it will need to be explicitly added.
Once certified, the calibration entries will exist in a gen3 calibration collection (and should be pushed into FITS headers as appropriate).  This collection will be indexed/have the VALID range used in the Gen3 dimensions.  I expect the way this will work is that there will be a way to do the same kind of query that is now done by pipetask: for a raw dataId and list of collections, which set of calibrations (or which explicit calibration) matches.
If we need to replace a certified calibration (due to a mistake, algorithm improvement, etc), then we'll do something to archive the previous calibration collection, and begin processing with a new validated and certified collection of calibrations.  I think this suggests adding additional dates to the collection, indicating the date range in which that collection was actually used.